### ISSUE TITLE
Cannot spy on browser Notification constructor function

### ISSUE THREAD

**Issue by bahmutov** [cypress-io/cypress#6129]
Cypress v3.8.1
Reproducible example in https://github.com/cypress-io/cypress-test-tiny/tree/spy-on-constructor

This test breaks with the error below

```js
it('spies on Notification', () => {
  cy.spy(window, 'Notification').as('Notification')
  const f = new Notification('Hello')
})
```

<img width="451" alt="Screen Shot 2020-01-09 at 10 56 44 AM" src="https://user-images.githubusercontent.com/2212006/72082888-e8e20600-32ce-11ea-9bb5-9835bea66280.png">

Note that spying on "regular" constructor functions works

```js
function Foo () {
  if (!new.target) {
    console.log('Foo was called without new!')
    return
  }
  console.log('inside Foo constructor')
}
window.Foo = Foo

it('stubs the constructor', () => {
  cy.stub(window, 'Foo')
  const f = new window.Foo()
  // new automatically converts whatever stub returns into an empty object
  expect(f).to.deep.equal({})
})
```

and if we wrap Notification with our function that calls `new` then notifications are working

```js
// workaround - spy on our wrapper of Notification
it('spies on Notification wrapper', () => {
  window.BrowserNotification = (text) => {
    return new Notification(text)
  }
  cy.spy(window, 'BrowserNotification').as('Notification')
  // actual notification will pop up (if Cypress notifications are enabled)
  const f = new BrowserNotification('Hello')
  cy.get('@Notification').should('have.been.calledOnce')
    .and('calledWithNew')
})
```
<img width="1280" alt="Screen Shot 2020-01-09 at 10 56 24 AM" src="https://user-images.githubusercontent.com/2212006/72082999-1a5ad180-32cf-11ea-83d5-36355eb1a53b.png">

So this seems like sinon.js problem in this case, but still very weird behavior

**Comment by bahmutov** [comment]
The place in Sinon.js in our code that calls the function it spies on

<img width="1283" alt="Screen Shot 2020-01-09 at 10 48 42 AM" src="https://user-images.githubusercontent.com/2212006/72083083-39596380-32cf-11ea-9307-e20ed3353ac3.png">

**Comment by jennifer-shehane** [comment]
This is fixed in our 4.0 Cypress branch, which includes a Sinon upgrade. https://github.com/cypress-io/cypress/pull/4226

<img width="558" alt="Screen Shot 2020-01-13 at 12 44 31 PM" src="https://user-images.githubusercontent.com/1271364/72235803-85274800-3602-11ea-9134-fbcd55a46e40.png">

@bahmutov Is it safe to close this as part of that release?

**Comment by bahmutov** [comment]
can this test be one of e2e tests in 4.0 branch?
```js
it('spies on Notification', () => {
  cy.spy(window, 'Notification').as('Notification')
  const f = new Notification('Hello')
  cy.get('@Notification').should('have.been.calledWith', 'Hello')
})
```
Then I will feel safe :)

**Comment by jennifer-shehane** [comment]
Hmm, I'm a bit unsure of where a test like this should go. @chrisbreiding Do we have any similar tests like this in our repo? Any idea where a test like this would go?

**Comment by chrisbreiding** [comment]
I think it could go along with the [other cy.spy() tests](https://github.com/cypress-io/cypress/blob/develop/packages/driver/test/cypress/integration/commands/agents_spec.coffee#L456) in the driver.

**Comment by flotwig** [comment]
Found the original Sinon issue that was fixed in 4.1.3: sinonjs/sinon#1265

**Comment by flotwig** [comment]
So... the `spy`ing works, but when Cypress goes to serialize the spy for printing, Sinon has a bug that causes it to throw an error. :smile: 

I opened a bug report here: https://github.com/sinonjs/sinon/issues/2215 Might open a PR, since it looks like a simple fix.

**Comment by flotwig** [comment]
Opened a PR over there to address the issue: sinonjs/sinon#2215

Once that gets merged and we update `sinon`, spying on constructors using `cy.spy` will be gravy. For now, it will only work via `Cypress.sinon.spy`, since that prevents the `toString` call.

**Comment by flotwig** [comment]
Sinon will release the fix in 9.0.0. Not sure why this issue has been left open... D:
